### PR TITLE
More info when database is missing a node, plus related pruning bugfix

### PR DESCRIPTION
**Pull request by carver**
### What was wrong?

When a node in the trie was missing, we:
1. got relatively little information back about what happened
2. would prune nodes *before* attempting to set/delete, so extra trie elements would disappear

### How was it fixed?

- Add a custom exception for a missing trie node
- Include information in the exception like the root hash, the key requested, and the hash of the node missing
- Add tests that uncovered pruning bug
- Fix pruning bug

TODO
- [x] Debug what appears to be situations where pruning is incomplete
- [x] Remove type check from custom exception (or maybe improve the error message and add it to all values?)

#### Cute Animal Picture

Cute animal picture

### CODE REVIEW

**Comment by carver** [comment]
Interesting, I wasn't sure circle was going to run on a draft

File: trie/hexary.py [hunk 1]
```diff
@@ -81,62 +85,95 @@ def _get(self, node, trie_key):
...
+
         self._set_root_node(new_node)
 
     def _set(self, node, trie_key, value):
-        self._prune_node(node)
-
         node_type = get_node_type(node)
 
+        # node is mutable, so capture the key for later pruning now
+        prune_key = self._get_prune_key(node)
```
Comment (pipermerriam) [inline trie/hexary.py]: This seems like potential for forgetting, what about something like:

```python
with self._prune_node(node):
    ...  # do things that might raise errors
```

The pruning would only happen *if* the context exits without exception.

**Review by pipermerriam — APPROVED** [review]

File: trie/hexary.py (hunk 1 above)
Comment (carver) [reply trie/hexary.py]: Argh, how did I miss that. Thanks!

File: trie/hexary.py [hunk 2]
```diff
@@ -248,6 +230,30 @@ def root_node(self, value):
...
+            node_body = None
+
+        # Nothing to prune
+        if node_body is None:
+            yield
+        else:
+            # Prune only if no exception is raised
+            try:
+                yield
+            except:
```
Comment (pipermerriam) [inline trie/hexary.py:252]: I don't think you need this try/except block at all (and we probably end up taking a performance hit for it).

```python
try:
    thing()
except:
    raise
else:
    something()

# equivalent to
thing()
something()
```

**Review by pipermerriam — COMMENTED** [review]
probably should be cleaned up.

File: trie/hexary.py (hunk 2 above)
Comment (carver) [reply trie/hexary.py:252]: #86

**Pull request merged**